contracts/PatriciaTree/Data.sol

```diff
@@ -0,0 +1,163 @@
+    bytes32 root;
+    Data.Edge rootEdge;
+    mapping(bytes32 => Data.Node) nodes;
+  }
+
+  // Returns a label containing the longest common prefix of `self` and `label`,
+  // and a label consisting of the remaining part of `label`.
+  function splitCommonPrefix(Label memory self, Label memory other) internal pure returns (
+    Label memory prefix,
+    Label memory labelSuffix
+  )
+  {
+    return splitAt(self, commonPrefix(self, other));
+  }
+
+  // Splits the label at the given position and returns prefix and suffix,
+  // i.e. 'prefix.length == pos' and 'prefix.data . suffix.data == l.data'.
+  function splitAt(Label memory self, uint pos) internal pure returns (Label memory prefix, Label memory suffix) {
+    assert(pos <= self.length && pos <= 256);
+    prefix.length = pos;
+    if (pos == 0) {
+      prefix.data = bytes32(0);
+    } else {
+      prefix.data = bytes32(uint(self.data) & ~uint(1) << 255 - pos);
+    }
+    suffix.length = self.length - pos;
+    suffix.data = self.data << pos;
+  }
+
+  // Returns the length of the longest common prefix of the two labels.
+  /*
+  function commonPrefix(Label memory self, Label memory other) internal pure returns (uint prefix) {
+    uint length = self.length < other.length ? self.length : other.length;
+    // TODO: This could actually use a "highestBitSet" helper
+    uint diff = uint(self.data ^ other.data);
+    uint mask = uint(1) << 255;
+    for (; prefix < length; prefix++) {
+      if ((mask & diff) != 0) {
+        break;
+      }
+      diff += diff;
+    }
+  }
+  */
+
+  function commonPrefix(Label memory self, Label memory other) internal pure returns (uint prefix) {
+    uint length = self.length < other.length ? self.length : other.length;
+    if (length == 0) {
+      return 0;
+    }
+    uint diff = uint(self.data ^ other.data) & ~uint(0) << 256 - length; // TODO Mask should not be needed.
+    if (diff == 0) {
+      return length;
+    }
+    return 255 - Bits.highestBitSet(diff);
+  }
+
+  // Returns the result of removing a prefix of length `prefix` bits from the
+  // given label (shifting its data to the left).
+  function removePrefix(Label memory self, uint prefix) internal pure returns (Label memory r) {
+    require(prefix <= self.length);
+    r.length = self.length - prefix;
+    r.data = self.data << prefix;
+  }
+
+  // Removes the first bit from a label and returns the bit and a
+  // label containing the rest of the label (shifted to the left).
+  function chopFirstBit(Label memory self) internal pure returns (uint firstBit, Label memory tail) {
+    require(self.length > 0);
+    return (uint(self.data >> 255), Label(self.data << 1, self.length - 1));
+  }
+
+  function edgeHash(Data.Edge memory self) internal pure returns (bytes32) {
+    return keccak256(self.node, self.label.length, self.label.data);
+  }
+
+  // Returns the hash of the encoding of a node.
+  function hash(Data.Node memory self) internal pure returns (bytes32) {
```

Comment:
`hash` is not a reserved keyword in Solidity to my surprise but still I'd prefer to steer away from using it. 
Let's rename this to `nodeEncodingHash` or something that makes tracing the function usage easier. There's also another usage of `hash` as a `Submission` struct member if you agree we should not have it at all.